### PR TITLE
Fix tests for macOS

### LINKED ISSUE #207: Tests fail on Mac OS
#### Describe the bug

Running on Mac OS 13.2.1 and Python 3.11, several test cases fail. While I expect the plotting tests to fail, also some functional tests fail. I need to further investigate this but I'm having issues with attaching the debugger on Mac OS.

#### Code sample

```zsh
poetry run pytest
```

For full output, click details below:
<details>

```zsh
(pyvrp-py3.11) wouterkool@Wouters-MacBook-Pro PyVRP % pytest
========================================================================== test session starts ===========================================================================
platform darwin -- Python 3.11.2, pytest-7.2.2, pluggy-1.0.0
rootdir: /Users/wouterkool/dev/PyVRP, configfile: pyproject.toml, testpaths: pyvrp
plugins: cov-4.0.0, anyio-3.6.2
collected 254 items                                                                                                                                                      

pyvrp/crossover/tests/test_selective_route_exchange.py .............                                                                                               [  5%]
pyvrp/diversity/tests/test_broken_pairs_distance.py .                                                                                                              [  5%]
pyvrp/educate/tests/test_Exchange.py ..................F                                                                                                           [ 12%]
pyvrp/educate/tests/test_LocalSearch.py ............                                                                                                               [ 17%]
pyvrp/educate/tests/test_MoveTwoClientsReversed.py F........                                                                                                       [ 21%]
pyvrp/educate/tests/test_RelocateStar.py ....                                                                                                                      [ 22%]
pyvrp/educate/tests/test_SwapStar.py .........                                                                                                                     [ 26%]
pyvrp/educate/tests/test_TwoOpt.py .........                                                                                                                       [ 29%]
pyvrp/educate/tests/test_neighbourhood.py ........                                                                                                                 [ 33%]
pyvrp/plotting/tests/test_plotting.py ..FF.                                                                                                                        [ 35%]
pyvrp/stop/tests/test_MaxIterations.py ........                                                                                                                    [ 38%]
pyvrp/stop/tests/test_MaxRuntime.py ............                                                                                                                   [ 42%]
pyvrp/stop/tests/test_NoImprovement.py ...........                                                                                                                 [ 47%]
pyvrp/stop/tests/test_TimedNoImprovement.py ................                                                                                                       [ 53%]
pyvrp/tests/test_CostEvaluator.py .........                                                                                                                        [ 57%]
pyvrp/tests/test_GeneticAlgorithm.py ................                                                                                                              [ 63%]
pyvrp/tests/test_Individual.py ............                                                                                                                        [ 68%]
pyvrp/tests/test_Matrix.py ....                                                                                                                                    [ 69%]
pyvrp/tests/test_PenaltyManager.py .............                                                                                                                   [ 74%]
pyvrp/tests/test_Population.py ..................F.                                                                                                                [ 82%]
pyvrp/tests/test_ProblemData.py .                                                                                                                                  [ 83%]
pyvrp/tests/test_Result.py .........                                                                                                                               [ 86%]
pyvrp/tests/test_Statistics.py .                                                                                                                                   [ 87%]
pyvrp/tests/test_SubPopulation.py ......                                                                                                                           [ 89%]
pyvrp/tests/test_TimeWindowSegment.py .....                                                                                                                        [ 91%]
pyvrp/tests/test_XorShift128.py ......                                                                                                                             [ 93%]
pyvrp/tests/test_read.py ................                                                                                                                          [100%]

================================================================================ FAILURES ================================================================================
_________________________________________________________________ test_relocate_after_depot_should_work __________________________________________________________________

    def test_relocate_after_depot_should_work():
        """
        This test exercises the bug identified in issue #142, involving a relocate
        action that should insert directly after the depot.
        """
        data = read("data/OkSmall.txt")
        cost_evaluator = CostEvaluator(20, 6)
        rng = XorShift128(seed=42)
    
        # This is a non-empty neighbourhood (so LS does not complain), but the only
        # client moves allowed by it will not improve the initial solution created
        # below. So the only improvements (1, 0)-exchange can make must come from
        # moving clients behind the depot of a route.
        neighbours = [[] for _ in range(data.num_clients + 1)]
        neighbours[2].append(1)
    
        ls = LocalSearch(data, rng, neighbours)
        ls.add_node_operator(Exchange10(data))
    
        # This individual can be improved by moving 3 into its own route, that is,
        # inserting it after the depot of an empty route. Before the bug was fixed,
        # (1, 0)-exchange never performed this move.
        individual = Individual(data, [[1, 2, 3], [4]])
        expected = Individual(data, [[1, 2], [3], [4]])
>       assert_equal(ls.search(individual, cost_evaluator), expected)
E       AssertionError: 
E       Items are not equal:
E        ACTUAL: <pyvrp._Individual.Individual object at 0x116373330>
E        DESIRED: <pyvrp._Individual.Individual object at 0x1163738b0>

pyvrp/educate/tests/test_Exchange.py:208: AssertionError
_______________________________________________________________________ test_single_route_OkSmall ________________________________________________________________________

    def test_single_route_OkSmall():
        """
        This test checks that MoveTwoClientsReversed properly solves the small
        instance where we know what is going on.
        """
        data = read("data/OkSmall.txt")
        cost_evaluator = CostEvaluator(20, 6)
        rng = XorShift128(seed=42)
    
        nb_params = NeighbourhoodParams(nb_granular=data.num_clients)
        ls = LocalSearch(data, rng, compute_neighbours(data, nb_params))
        ls.add_node_operator(MoveTwoClientsReversed(data))
    
        individual = Individual(data, [[1, 4, 2, 3]])
        improved_individual = ls.search(individual, cost_evaluator)
    
        # The new solution should strictly improve on our original solution.
        assert_equal(improved_individual.num_routes(), 1)
        current_cost = cost_evaluator.penalised_cost(individual)
        improved_cost = cost_evaluator.penalised_cost(improved_individual)
        assert_(improved_cost < current_cost)
    
        # (2, 3) was inserted after 1 as 1 -> 3 -> 2 -> 4. Then (1, 3) got inserted
        # after 2 as 2 -> 3 -> 1 -> 4.
>       assert_equal(improved_individual.get_routes(), [[2, 3, 1, 4], [], []])
E       AssertionError: 
E       Items are not equal:
E       item=0
E       item=0
E       
E        ACTUAL: 4
E        DESIRED: 2

pyvrp/educate/tests/test_MoveTwoClientsReversed.py:38: AssertionError
_________________________________________________________________________ test_plot_result[png] __________________________________________________________________________

args = (), kwds = {'extension': 'png', 'request': <FixtureRequest for <Function test_plot_result[png]>>}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 6.948):
E               result_images/test_plotting/plot_result.png
E               result_images/test_plotting/plot_result-expected.png
E               result_images/test_plotting/plot_result-failed-diff.png

/usr/local/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py:81: ImageComparisonFailure
________________________________________________________________________ test_plot_instance[png] _________________________________________________________________________

args = (), kwds = {'extension': 'png', 'request': <FixtureRequest for <Function test_plot_instance[png]>>}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 6.837):
E               result_images/test_plotting/plot_instance.png
E               result_images/test_plotting/plot_instance-expected.png
E               result_images/test_plotting/plot_instance-failed-diff.png

/usr/local/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py:81: ImageComparisonFailure
_____________________________________________________________________ test_purge_removes_duplicates ______________________________________________________________________

    def test_purge_removes_duplicates():
        data = read("data/RC208.txt", "solomon", "dimacs")
        cost_evaluator = CostEvaluator(20, 6)
        params = PopulationParams(min_pop_size=20, generation_size=5)
        rng = XorShift128(seed=42)
    
        pop = Population(bpd, params=params)
    
        for indiv in make_random_solutions(params.min_pop_size, data, rng):
            pop.add(indiv, cost_evaluator)
    
        assert_equal(len(pop), params.min_pop_size)
    
        # This is the individual we are going to add a few times. That should make
        # sure the relevant subpopulation definitely contains duplicates.
        individual = Individual.make_random(data, rng)
        assert_(not individual.is_feasible())
    
        for _ in range(params.generation_size):
            pop.add(individual, cost_evaluator)
    
        # Make sure we have not yet purged, and increase the minimum population
        # size by one to make sure we're definitely not removing *all* of the
        # duplicate individuals.
        assert_(pop.num_infeasible() != params.min_pop_size)
        params.min_pop_size += 1
    
        # Keep adding individuals until we have had a purge, and returned to the
        # minimum population size.
        while pop.num_infeasible() != params.min_pop_size:
            pop.add(Individual.make_random(data, rng), cost_evaluator)
    
        # Since duplicates are purged first, there should now be only one of them
        # in the subpopulation. There cannot be zero, because we made sure of that.
        duplicates = 0
        for other in pop:
            if other == individual:
                duplicates += 1
    
>       assert_equal(duplicates, 1)
E       AssertionError: 
E       Items are not equal:
E        ACTUAL: 0
E        DESIRED: 1

pyvrp/tests/test_Population.py:319: AssertionError

---------- coverage: platform darwin, python 3.11.2-final-0 ----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
pyvrp/GeneticAlgorithm.py                       101     10    90%
pyvrp/PenaltyManager.py                          59      0   100%
pyvrp/Population.py                              61      0   100%
pyvrp/Result.py                                  31      0   100%
pyvrp/Statistics.py                              78      0   100%
pyvrp/__init__.py                                11      0   100%
pyvrp/crossover/__init__.py                       1      0   100%
pyvrp/crossover/selective_route_exchange.py      13      0   100%
pyvrp/diagnostics/__init__.py                     1      0   100%
pyvrp/diagnostics/get_route_statistics.py        45     26    42%
pyvrp/diversity/__init__.py                       1      0   100%
pyvrp/educate/LocalSearch.py                     34      7    79%
pyvrp/educate/__init__.py                         9      0   100%
pyvrp/educate/neighbourhood.py                   44      0   100%
pyvrp/exceptions.py                               2      0   100%
pyvrp/plotting/__init__.py                       10      0   100%
pyvrp/plotting/plot_coordinates.py               17      1    94%
pyvrp/plotting/plot_demands.py                   16      1    94%
pyvrp/plotting/plot_diversity.py                 19      1    95%
pyvrp/plotting/plot_instance.py                  14      0   100%
pyvrp/plotting/plot_objectives.py                29      1    97%
pyvrp/plotting/plot_result.py                    18      0   100%
pyvrp/plotting/plot_route_schedule.py            79      4    95%
pyvrp/plotting/plot_runtimes.py                  19      1    95%
pyvrp/plotting/plot_solution.py                  26      0   100%
pyvrp/plotting/plot_time_windows.py              18      1    94%
pyvrp/read.py                                    62      3    95%
pyvrp/stop/MaxIterations.py                       9      0   100%
pyvrp/stop/MaxRuntime.py                         12      0   100%
pyvrp/stop/NoImprovement.py                      14      0   100%
pyvrp/stop/StoppingCriterion.py                   1      0   100%
pyvrp/stop/TimedNoImprovement.py                  8      0   100%
pyvrp/stop/__init__.py                            5      0   100%
-----------------------------------------------------------------
TOTAL                                           867     56    94%
Coverage XML written to file coverage.xml

======================================================================== short test summary info =========================================================================
FAILED pyvrp/educate/tests/test_Exchange.py::test_relocate_after_depot_should_work - AssertionError: 
FAILED pyvrp/educate/tests/test_MoveTwoClientsReversed.py::test_single_route_OkSmall - AssertionError: 
FAILED pyvrp/plotting/tests/test_plotting.py::test_plot_result[png] - matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 6.948):
FAILED pyvrp/plotting/tests/test_plotting.py::test_plot_instance[png] - matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 6.837):
FAILED pyvrp/tests/test_Population.py::test_purge_removes_duplicates - AssertionError: 
===================================================================== 5 failed, 249 passed in 15.38s =====================================================================
```

</details>

#### Expected Output

Expect all of the tests to succeed, except for the plotting tests for which it would be nice if we could have a more robust solution.

### PR DESCRIPTION
**Pull request by leonlan**
This PR:

- [x] Closes #207.
- [x] Adds and passes relevant tests.
- [ ] Has well-formatted code and documentation.

<details>

**Notes**:

- It is essential that you add a test when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

File: pyvrp/search/tests/test_Exchange.py
Comment (leonlan) [inline pyvrp/search/tests/test_Exchange.py:1]: Changing the seed from 42 to 1 makes this test work on macOS.

File: pyvrp/plotting/tests/test_plotting.py
Comment (leonlan) [inline pyvrp/plotting/tests/test_plotting.py:1]: See below for diff details. Increasing the tolerance helps to fix this (for me at least). It's not a robust solution, but the tolerance doesn't really seem very high to me. I compared a result figure with no iterations which resulting in an RMSE of 18.

<details>
Plot instance diff:
plot_instance-failed-diff

Plot result diff:
plot_result-failed-diff
</details>

Comment (N-Wouda) [reply pyvrp/plotting/tests/test_plotting.py:1]: I don't know if a robust solution exists. Plotting is just a bit flaky, I guess. This seems to work well enough.

**Review by N-Wouda — APPROVED** [review]

**Comment by N-Wouda** [comment]
Do all tests work again on OS X with this PR?

**Comment by leonlan** [comment]
> Do all tests work again on OS X with this PR? I'm assuming yes, but I have no way to verify easily.

The MoveTwoClientsReversed test definitely should work now since the test is deterministic. The plotting tests work for my macbook, but I'm not 100% sure whether it works for other macOS systems. @wouterkool could you check if this works for you?

**Comment by N-Wouda** [comment]
We might be able to add a windows and mac os job to the CI. Just one, not the whole array, to see if they also work. I'll have a look at that sometime soon.

**Pull request merged**